### ISSUE TITLE
subscription keeps receiving messages after subscription.close() was called

### ISSUE THREAD

**Issue by ctavan** [googleapis/nodejs-pubsub#66]
#### Environment details

  - OS: Kubernetes Engine
  - Node.js version: 9.3.0
  - npm version: –
  - @google-cloud/pubsub version: 0.16.2

#### Steps to reproduce

During my investigations on #62 I cluttered the library code with `console.log()` statements ([see this branch for reference](https://github.com/googleapis/nodejs-pubsub/compare/master...contentpass:fix-ack-on-subscription-close-debug)) and found an interesting pattern:

After `subscription.close()` is called, the subscription can still receive many new messages, that all need to be `nack`ed. For the investigation [I've placed a debug statement in the `modifyAckDeadline_()` method](https://github.com/googleapis/nodejs-pubsub/compare/master...contentpass:fix-ack-on-subscription-close-debug#diff-3d29c4447546c72118ed5d5cbf38ab8bR954).

Here's the output of this log statement:

```
{
  "message": "modifyAckDeadline_ non-streaming",
  "acknowledgeInstanceId": "230edb80-718a-4b0a-8e9f-083da5da3fd2",
  "deadline": 0,
  "count": 2202,
  "bytes": 359042
}
{
  "message": "modifyAckDeadline_ non-streaming",
  "acknowledgeInstanceId": "3d749b9f-0156-4120-aee5-fa9c1ce455b0",
  "deadline": 0,
  "count": 5068,
  "bytes": 826668
}
{
  "message": "modifyAckDeadline_ non-streaming",
  "acknowledgeInstanceId": "2b2b66b9-723b-4cb6-a654-8ebad6ff298a",
  "deadline": 0,
  "count": 7889,
  "bytes": 1287063
}
{
  "message": "modifyAckDeadline_ non-streaming",
  "acknowledgeInstanceId": "c7b13ac3-5440-438e-9eac-f6bda0f59bc6",
  "deadline": 0,
  "count": 9270,
  "bytes": 1512777
}
{
  "message": "modifyAckDeadline_ non-streaming",
  "acknowledgeInstanceId": "aa3de1e8-55bd-432f-b9b0-8b1c057a9b76",
  "deadline": 0,
  "count": 11806,
  "bytes": 1926374
}
{
  "message": "modifyAckDeadline_ non-streaming",
  "acknowledgeInstanceId": "eed24ea6-2bce-4137-8ad8-c49dc16e528b",
  "deadline": 0,
  "count": 280,
  "bytes": 45585
}
{
  "message": "modifyAckDeadline_ non-streaming",
  "acknowledgeInstanceId": "9c8653a5-2117-4454-9e96-fd154e75bdf0",
  "deadline": 0,
  "count": 2861,
  "bytes": 466398
}
```

Or looking just at the ackId-counts:

```
  "count": 2202,
  "count": 5068,
  "count": 7889,
  "count": 9270,
  "count": 11806,
  "count": 280,
  "count": 2861,
```

As you can see the counts keep on increasing for a while before the eventually drop. Please note that this all happens _after_ the `subscription.close()` call.

Would it somehow be possible to stop receiving new messages as soon as `.close()` has been called? Or is the observed behavior intended?

**Comment by callmehiphop** [comment]
Interesting, about how long would you say this occurs for after closing a subscription?

@murgatroid99 would there be any reason why gRPC would continue to deliver messages after `end()` has been called on a duplex stream?

**Comment by ctavan** [comment]
@callmehiphop I have just checked the logs again and this happened for up to 10 seconds after the `subscription.close()` call.

Btw, where in https://github.com/googleapis/nodejs-pubsub/blob/ace59b086c9f72fb1e582398f7cf232b1428b6bc/src/subscription.js#L381-L395 does the `end()` happen?

**Comment by murgatroid99** [comment]
Those streams are normal Node streams, where calling the `end` method indicates that you are done writing to the stream. But in general, saying that you are done writing to the stream does not mean that you do not want to read more data from it, and as long as the server keeps sending messages, gRPC will keep outputting them. If you want to stop it abruptly, you can call the `cancel` method, but keep in mind that that will end the call with a non-OK status, which may be propagated up to the user as an error.

**Comment by callmehiphop** [comment]
@ctavan we've got a fix prepared for this (#92) if you have the bandwidth to test it out that would be fantastic! Otherwise we should have it released within the next day or two.
